contrib/bloop/src/mill.contrib.bloop/BloopImpl.scala

```diff
@@ -0,0 +1,339 @@
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Scalac
+    ////////////////////////////////////////////////////////////////////////////
+
+    val scalaConfig = module match {
+      case s: ScalaModule =>
+        val semanticDb = s.resolveDeps(s.scalaVersion.map {
+          case scalaV if semanticDBSupported(scalaV) =>
+            Agg(ivy"org.scalameta:semanticdb-scalac_$scalaV:$semanticDBVersion")
+          case _ => Agg()
+        })
+
+        T.task {
+          val pluginCp = semanticDb() ++ s.scalacPluginClasspath()
+          val pluginOptions = pluginCp.map { pathRef =>
+            s"-Xplugin:${pathRef.path}"
+          }
+
+          val allScalacOptions =
+            (s.scalacOptions() ++ pluginOptions ++ semanticDBOptions).toList
+          Some(
+            BloopConfig.Scala(
+              organization = "org.scala-lang",
+              name = "scala-compiler",
+              version = s.scalaVersion(),
+              options = allScalacOptions,
+              jars = s.scalaCompilerClasspath().map(_.path.toNIO).toList,
+              analysis = None,
+              setup = None
+            )
+          )
+        }
+      case _ => T.task(None)
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Platform (Jvm/Js/Native)
+    ////////////////////////////////////////////////////////////////////////////
+
+    val platform = T.task {
+      BloopConfig.Platform.Jvm(
+        BloopConfig.JvmConfig(
+          home = T.ctx().env.get("JAVA_HOME").map(s => Path(s).toNIO),
+          options = module.forkArgs().toList
+        ),
+        mainClass = module.mainClass()
+      )
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Tests
+    ////////////////////////////////////////////////////////////////////////////
+
+    val testConfig = module match {
+      case m: TestModule =>
+        T.task {
+          Some(
+            BloopConfig.Test(
+              frameworks = m
+                .testFrameworks()
+                .map(f => Config.TestFramework(List(f)))
+                .toList,
+              options = Config.TestOptions(
+                excludes = List(),
+                arguments = List()
+              )
+            )
+          )
+        }
+      case _ => T.task(None)
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    //  Ivy dependencies + sources
+    ////////////////////////////////////////////////////////////////////////////
+
+    val scalaLibraryIvyDeps = module match {
+      case x: ScalaModule => x.scalaLibraryIvyDeps
+      case _              => T.task { Loose.Agg.empty[Dep] }
+    }
+
+    /**
+      * Resolves artifacts using coursier and creates the corresponding
+      * bloop config.
+      */
+    def artifacts(repos: Seq[coursier.Repository],
```

Comment:
Calling coursier directly as the logic needs to retain the `dep => path` relation